### PR TITLE
NEW: Enable MPI for lstsq_grad method

### PR DESCRIPTION
**Pull request by xiaodong-yu**
## Purpose

Add multi-node to lstsq_grad method by using mpi4py.
## Approach

Apply MPI operations onto the communication and sync spots in the `lstsq_grad` method, in addition to the threadpool operations.
### Submitter
- [ ] Write a helpfully descriptive pull request title.
- [ ] Organize changes into logically grouped commits with descriptive commit messages.
- [ ] Document all new functions.
- [ ] Write tests for new functions or explain why they are not needed.
- [ ] Build the documentation successfully
- [ ] Use [`yapf`](https://github.com/google/yapf) to format python code.

### Reviewer
- [x] Actually read all of the code.
- [x] Run the new code yourself.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

**Comment by carterbox** [comment]
@xiaodong-yu, I pulled these images from `tests/results/lstsq_grad` Can you see that the texture of the noise switches half way down the image with 2 processes?
| 1 process | 2 processes |
|-|-|
| 0-phase-mpi-1 | 0-phase-mpi-2 |

Something about this implementation is causing this. Any ideas? Do you want a large dataset to test?

**Review by carterbox — CHANGES_REQUESTED** [review]
@xiaodong-yu, I pulled these images from `tests/results/lstsq_grad` Can you see that the texture of the noise switches half way down the image with 2 processes?
| 1 process | 2 processes |
|-|-|
| 0-phase-mpi-1 | 0-phase-mpi-2 |

Something about this implementation is causing this. Any ideas? Do you want a large dataset to test?

**Comment by xiaodong-yu** [comment]
> @xiaodong-yu, I pulled these images from `tests/results/lstsq_grad` Can you see that the texture of the noise switches half way down the image with 2 processes?
> 
> 1 process	2 processes
> 0-phase-mpi-1	0-phase-mpi-2
> Something about this implementation is causing this. Any ideas? Do you want a large dataset to test?

@carterbox, I can't clearly see any differences between these two images. Could you please explain it a little bit more to me? Also, is the 1 process image reconstructed by multiple GPUs? I highly suspect that any differences come from the mean issue you mentioned before. I find that the splitting of the test data is not even.

**Comment by carterbox** [comment]
It's more obvious in this image (the amplitude of the first view reconstructed by the lstsq test). I increased the problem size and increased the number of processes to 4. I'm using a single GPU because the hardware layout shouldn't affect the quality.

0-ampli

You can see there are 4 distinct horizontal stripe regions in the reconstruction. Probably the common probe constraint isn't correctly implemented.

**Comment by xiaodong-yu** [comment]
> It's more obvious in this image (the amplitude of the first view reconstructed by the lstsq test). I increased the problem size and increased the number of processes to 4. I'm using a single GPU because the hardware layout shouldn't affect the quality.
> 
> 0-ampli
> 
> You can see there are 4 distinct horizontal stripe regions in the reconstruction. Probably the common probe constraint isn't correctly implemented.

OK, now I see the problem. I agree the hardware layout won't affect the results. But I don't think the issue comes from this MPI implementation since it doesn't really change the parallelization design. Any algorithmic problem very likely is caused by the multi-device implementation, even within a single node. I'll check the previous PR to see if it has the same issue with single node multi-GPU.

**Comment by xiaodong-yu** [comment]
@carterbox  I revised the code and don't see the issue anymore with the test image. Please have a review and let me know if you still see the problem in your reconstructions.

**Comment by carterbox** [comment]
I plan to look at this on Thursday or Friday.

**Comment by carterbox** [comment]
@xiaodong-yu, the stripes still appears for me. I have pushed a commit which changes the ptycho tests so we are looking at the same size image. I run the following test and there is a noticeable line across the middle of the image.

`mpirun -n 2 pytest tests/test_ptycho.py -vs -k 'lstsq and not variable'`

Using more than two mpi processes causes an exception in mpi4py.

**Comment by xiaodong-yu** [comment]
> @xiaodong-yu, the stripes still appears for me. I have pushed a commit which changes the ptycho tests so we are looking at the same size image. I run the following test and there is a noticeable line across the middle of the image.
> 
> `mpirun -n 2 pytest tests/test_ptycho.py -vs -k 'lstsq and not variable'`
> 
> Using more than two mpi processes causes an exception in mpi4py.

I still didn't see the stripe region in the image (2 processes, as attached). Did you refer to the phase images in `tests/result/lstsq_grad`?

1-phase

**Comment by carterbox** [comment]
No, I'm looking at the amplitude image.

**Comment by xiaodong-yu** [comment]
@carterbox Why the amplitude images I found in the folder are the following random images?

1-ampli

**Comment by carterbox** [comment]
There are two object views and two probes reconstructed, so there should be 8 images:
```
tests/result/lstsq_grad/0-ampli.png
tests/result/lstsq_grad/0-phase.png
tests/result/lstsq_grad/0-probe-ampli.png
tests/result/lstsq_grad/0-probe-phase.png
tests/result/lstsq_grad/1-ampli.png
tests/result/lstsq_grad/1-phase.png
tests/result/lstsq_grad/1-probe-ampli.png
tests/result/lstsq_grad/1-probe-phase.png
```
Image 0 has amplitude "satyre". Image 1 has amplitude "coins". Probe 0 has phase "cryptomeria". Probe 1 has phase "bombus".

**Comment by xiaodong-yu** [comment]
@carterbox, the issue is solved. Please try it again.

File: src/tike/ptycho/ptycho.py [hunk 1]
```diff
@@ -309,19 +309,41 @@ def reconstruct(
...
+        n2 = np.sqrt(comm.reduce(n2, 'cpu'))
 
-    rescale = (np.linalg.norm(np.ravel(np.sqrt(data[..., i, :, :]))) /
-               np.linalg.norm(np.ravel(np.sqrt(intensity))))
+    rescale = n1 / n2
 
     logger.info("object and probe rescaled by %f", rescale)
 
-    probe *= rescale
+    probe[0] *= rescale
```
Comment (carterbox) [inline src/tike/ptycho/ptycho.py:344]: This only rescales the probe for the zeroth GPU? What happens when there are multiple GPUs on each node?

File: src/tike/ptycho/solvers/divided.py [hunk 2]
```diff
@@ -307,24 +356,48 @@ def _update_nearplane(op, comm, nearplane, psi, scan_, probe, unique_probe,
...
+                    dest='gpu',
+                )
+            else:
+                weighted_step_psi[0] = comm.pool.reduce_mean(
+                    weighted_step_psi,
+                    axis=-5,
+                )[..., 0, 0, 0]
+                common_grad_psi[0] = comm.reduce(common_grad_psi, 'gpu')
 
+            #test = cp.ones(common_grad_psi[0].shape, common_grad_psi[0].dtype)
```
Comment (carterbox) [inline src/tike/ptycho/solvers/divided.py:375]:
```suggestion
```

File: src/tike/ptycho/solvers/divided.py [hunk 3]
```diff
@@ -307,24 +356,48 @@ def _update_nearplane(op, comm, nearplane, psi, scan_, probe, unique_probe,
...
             psi[0] += weighted_step_psi[0] * common_grad_psi[0]
+            #print(psi[0].tolist())
+            #exit()
```
Comment (carterbox) [inline src/tike/ptycho/solvers/divided.py:378]:
```suggestion
```

File: src/tike/ptycho/solvers/divided.py [hunk 4]
```diff
@@ -307,24 +356,48 @@ def _update_nearplane(op, comm, nearplane, psi, scan_, probe, unique_probe,
...
+            if comm.use_mpi:
+                weighted_step_probe[0] = comm.Allreduce_mean(
+                    weighted_step_probe,
+                    axis=-5,
+                )
+                common_grad_probe[0] = comm.Allreduce_mean(
+                    common_grad_probe,
+                    axis=-5,
+                )
+                #print(weighted_step_probe[0])
```
Comment (carterbox) [inline src/tike/ptycho/solvers/divided.py:391]:
```suggestion
```

**Review by carterbox — CHANGES_REQUESTED** [review]
Looks correct when I test a single GPU with multiple process. However, I have concern about rescaling (described below).

File: src/tike/ptycho/ptycho.py (hunk 1 above)
Comment (xiaodong-yu) [reply src/tike/ptycho/ptycho.py:344]: In your original implementation, you give probe[0] as the input to _rescaling, so your probe is actually probe[0]. I'm trying to keep my parallelization consistent with the algorithm of original codes.

Comment (carterbox) [reply src/tike/ptycho/ptycho.py:344]: That's a good goal, but in this case, I think my original implementation was incorrect for multi-gpu. If we don't rescale `probe[0:ngpu]`, then we will have the problem you just fixed (the probes not being the same for all nodes/devices). Do you agree?

Comment (carterbox) [reply src/tike/ptycho/ptycho.py:343]: I see that the broadcasting was happening in a place I didn't expect, so I moved it.

File: src/tike/ptycho/solvers/divided.py (hunk 2 above)
Comment (carterbox) [reply src/tike/ptycho/solvers/divided.py:375]: @xiaodong-yu, is there a reason you want to keep these commented-out line of code?

Comment (xiaodong-yu) [reply src/tike/ptycho/solvers/divided.py]: I overlooked this suggestion.

**Review by carterbox — APPROVED** [review]

**Pull request merged**